cranelift-codegen/src/redundant_reload_remover.rs

```diff
@@ -0,0 +1,462 @@
+// spill slot into a different register is still a win, since we can convert the fill into a
+// reg-reg copy.
+//
+// === Other tricks that don't seem to help
+//
+// Tracking the effects of InstructionData::Copy in the environment for the case when both source
+// and destination are registers.  But that didn't cause any more fills to be removed or converted
+// to copies.  It's not clear why.
+//
+// It's also possible for a spill to be detected as redundant, if such a binding already appears
+// in the environment.  This was also implemented, but it was almost entirely ineffective, so was
+// removed again.
+
+// This describes a spill slot in the obvious way.  It also tracks the SSA value name associated
+// with the slot.
+#[derive(Clone, Copy)]
+struct SlotInfo {
+    value: Value,
+    offset: i32,
+    size: u32,
+}
+
+// Find out if the range `[offset, +size)` overlaps with the range in `si`.
+fn any_overlap(si: &SlotInfo, offset: i32, size: u32) -> bool {
+    let a_offset = si.offset as i64;
+    let a_size = si.size as i64;
+    let b_offset = offset as i64;
+    let b_size = size as i64;
+    let no_overlap = a_offset + a_size <= b_offset || b_offset + b_size <= a_offset;
+    !no_overlap
+}
+
+// Find, in `reginfo`, the register bank that `reg` lives in, and return the lower limit and size
+// of the bank.  This is so the caller can conveniently iterate over all `RegUnits` in the bank
+// that `reg` lives in.
+fn find_bank_limits(reginfo: &RegInfo, reg: RegUnit) -> (RegUnit, u16) {
+    if let Some(bank) = reginfo.bank_containing_regunit(reg) {
+        return (bank.first_unit, bank.units);
+    }
+    // We should never get here, since `reg` must come from *some* RegBank.
+    panic!("find_regclass_limits: reg not found");
+}
+
+// The redundant-fill remover processes each EBB independently.  It makes a single forward pass
+// over each EBB, carrying along a running environment, `AvailEnv`, that maps `RegUnits` to an
+// `Option<SlotInfo>`.
+//
+// If the triplet `(regunit, offset, size)` appears in the environment, this indicates that, at
+// this point, `regunit` and the stack slot denoted by `(offset, size)` have the same value.  We
+// also keep track of the SSA `Value` associated with the `reg`, because that is necessary to
+// create `copy` instructions that read `reg`.
+//
+// Invariants for the environment:
+//
+// The environment may have multiple different registers bound to the same stack slot -- that is,
+// `(offset, size)` pair.  That's OK, and reflects the reality that those two registers contain
+// the same value.
+//
+// The `value` fields must however be unique: it isn't allowable to have two entries in the
+// environment that have the same `value` field.  To do so would be to imply that that SSA value
+// has two different locations, since each environment entry is associated with a different
+// register, and so this would imply that the Value resided in both of those registers.  Function
```

Comment:
I am not sure if this is technically true: we're operating after register allocations, which may introduce spurious moves that aren't tied to SSA semantics. Relaxing this invariant would mean that we could **not** have to delete the src binding in a register-to-register move. I agree that's it's useful to have such an invariant for clarity of understanding, but I am wondering if this is required to have it? (the SSA value is only used to transform fills with inexact bindings in reg -> reg2 copies)

Reply:
This is an interesting point.  Now that we've established the need for a copy_to_ssa instruction that copies from *any* register to "SSA-world", we could indeed relax this invariant.  However, I suspect it wouldn't give any improvement, because the (only?) reason to divert a value from one register to a different one is that we want to immediately overwrite the source with a new value.

Reply:
Actually, doing this removes about 0.5% more loads (statically counting) and doesn't seem to break anything.  So I'll enable it by default.